### DISCUSSION TITLE
useFormContext returns null

### DISCUSSION THREAD

**Discussion by evgvis** [react-hook-form/react-hook-form#3894] [Q&A] [answered]
Hi. i have some internal library where i created few components based on react-hook-form@5.7.2. everything was ok when i build library in one js bundle. then i have split library by component (1 component = 1 file) and forms components broke.
i see an error "Cannot destructure property 'control' of 'Object(...)(...)' as it is null" because useFormContext returns null;

some part of my components.
Form:
```
const Form = (props) => {
    const { formProps = {}, className = "", children } = props;

    const reactHookFormMethods = useForm({ validateCriteriaMode: "all", ...formProps });

    return (
        <FormContext {...reactHookFormMethods} >
            <form>
                {children}
                <Button .../>
            </form>
        </FormContext>
    );
};
```

Input:
```
const Input = (props) => {
    const { name, label, ... } = props;
    const { register, errors = {}, getValues, triggerValidation, setError, clearError } = useFormContext();
    
    const inputRef = useRef();

    return (
        <div>
            <label htmlFor={name}>
                <InputComponent
                    id={name}
                    name={name}
                    ref={e => {
                        register(e, validation);
                        inputRef.current = e;
                    }}
                />
                <span>{label}</span>
            </label>
        </div>
    )
}
```

**Comment by bluebill1049** [comment]
make sure your `FormProvider` does wrap up the entire forms.

**Reply by evgvis** [reply]
thank you, but all my form-components inside my Form component:

```
import Form from "ui-components/forms/Form";
import Input from "ui-components/forms/Input";

const Layout = (props) => {
	return (
		<Form formProps={{ onSaveFunc: console.log }} >
			<Input name="any" label="Enter anything" />
		</Form>
	(
};
```

maybe i have problem because of wrong config babel or webpack. what do you think ?

**Reply by bluebill1049** [reply]
Don't think so, you may want to reproduce that in a CSB, happy to take a look at it.

**Reply by evgvis** [reply]
I'd love to, but the library deployed to internal npm.

**Reply by evgvis** [reply]
ps. if you have some time, maybe you want to make small call where i will show you code of two my projects

**Reply by uttampun44** [reply]
@bluebill1049  You comment save my day.
`FormProvider`  fix my issue.

**Comment by evgvis** [comment]
i have updated react-hook-form. but with 6.14.1 i have the same error(

**Comment by evgvis** [comment]
@bluebill1049 hi)
i created test-lib with few component - https://github.com/evgvis/test-lib
then published it to npm - https://www.npmjs.com/package/@evgvis/test-lib
and created csb - https://codesandbox.io/s/xenodochial-sutherland-wo2hp?file=/src/App.js

may i ask you to review it ;) 
thanks

**Reply by bluebill1049** [reply]
where did you use `FormPrivder`?

**Reply by evgvis** [reply]
in [Form](https://github.com/evgvis/test-lib/blob/main/src/forms/Form.js) component

**Reply by bluebill1049** [reply]
yea, it's a strange one. not sure why this is happening.

**Reply by evgvis** [reply]
oh(( sad to hear this. thanks for trying to help ;)

**Reply by bluebill1049** [reply]
do you have a different instance of `React`?

**Reply by evgvis** [reply]
for now i use 16.13.1, should i try 17.0.1?

**Reply by bluebill1049** [reply]
that should be fine tho...

**Comment by evgvis** [comment, marked as answer]
@bluebill1049 i solved this issue)
basically what i need it is just adding react-hook-form as external library in [webpack config](https://github.com/evgvis/test-lib/blob/main/webpack.config.js)
and now, [csb](https://codesandbox.io/s/xenodochial-sutherland-wo2hp?file=/src/App.js) is working :)

**Reply by bluebill1049** [reply]
oh nice @evgvis well done mate. you sort it out.

**Reply by viniciusgandrade** [reply]
Hi @evgvis. Can you share the source code? The link in https://github.com/evgvis/test-lib/blob/main/webpack.config.js was broken. Thanks.

**Reply by evgvis** [reply]
@viniciusgandrade @Markila-BOT 
Sorry guys. the code was lost

**Reply by Moshyfawn** [reply]
I guess, you did something along the lines of:
```js
externals: {
  'react-hook-form': 'react-hook-form'
}
```
Refer to [externals section of Webpack docs](https://webpack.js.org/configuration/externals/#externals)

**Reply by evgvis** [reply]
@Markila-BOT try to add _react_ and _react-dom_ to _externals_.

**Reply by MrEfrem** [reply]
I have the same problem and only in webpack. When I build a library via Rollup and expose modules as ES modules then all works well. But Webpack doesn’t have full support of ES modules in the output yet and I have to use commonjs-static, although Webpack’s team writes this format is compatible with both commonjs and ES modules. I also added react, react-hook-form and all other dependencies from package.json in externals, but I still see this error. It’s really very weird because if look to  output bundle I see that all dependencies imported by require(""). Maybe because Webpack wraps all output source code in some webpack functions it works incorrectly. Would be good I somebody create a successful repo to see the solution.

**Reply by MerlinMason** [reply]
I have a similar problem... bundling a component library with Rollup (as CJS and ESM), publishing as npm package and then consuming into a NextJS app.

Looking at the bundled code I can see it creates the following
```js
const HookFormContext = React__default.createContext(null);
const useFormContext = () => React__default.useContext(HookFormContext);
```

Later in the code when `useFormContext()` is called I get an error as it returns null.

**Reply by MrEfrem** [reply]
I only give you a little hint. Check a bundle in library and nextjs. Both should import react-hook-form as CJS or ESM, because the form library contains different bundles for CJS and ESM. In my case library used CJS and App ESM bundle, so context was null.

**Reply by MerlinMason** [reply]
I'm using a Rollup plugin https://github.com/pmowrer/rollup-plugin-peer-deps-external which makes all my design systems peerDependencies `external` in the Rollup config.

In my NextJS app I've checked whats actually shipped from the design system, there's no instance of react-hook-form in the packages node_modules. So I assume when the packaged code runs `import { useFormContext } from 'react-hook-form';` it should be importing the same instance my NextJS code is using.

If I've understood what's been said here correctly this should work... but it doesn't :(

**Reply by MerlinMason** [reply]
Sorry... answering my own question here: it turns out it was nothing to do with packaging / webpack in my case but instead that `useFormContext()` returns null when server side rendering. Not sure if this is known/expected?

**Reply by smeng9** [reply]
Hi, I have experienced the same error, The reason is more related to how babel trans piles the createContext. Even when we import useFormContext using externals, two context will be created: one will be used for provider, the other does not have a provider and returns null. The best way to do is always separate the context and all the code just import one context. We need to pay attention to which of ESM or CJS are using and need to be consistent.

**Reply by boyboi86** [reply]
any updates? I'm facing the same issue, create-react-app (react 18) doesn't have webpack so I'm using react-app-rewired instead. so is there a way to resolve this?

**Reply by tiavina-mika** [reply]
Any solution? I'm using NextJs 12 with React 18 and TypeScript

**Reply by TkaczykAdam** [reply]
My not-perfect solution is to wrap the problematic components with  [NoSsr](https://mui.com/material-ui/react-no-ssr/).
 @bluebill1049 can you help us?

**Reply by mikita-kh** [reply]
I had the same issue with `vitejs`. In my case, the following solution has been done
```
export default defineConfig({
  resolve: {
    alias: {
      'react-hook-form': require.resolve('react-hook-form'),
    },
  },
```

As I understand, the main reason for this issue is that hooks `useForm` and `useFormContext` can be imported from different node_modules packages. In my case, `useForm` has been imported from `react-hook-form/dist/index.cjs.js` otherwise `useFormContext` has been imported from `react-hook-form/dist/index.esm.mjs`.

**Reply by flomillot** [reply]
Be careful to both compile your projects in ES6. If the library compile into ES5 and the main project into ES6, then the first one will import react-hook-form via the cjs file (CommonJS) and the other via the esm file (ES Modules).
We had to upgrade our old project (the library) into ES6. Maybe it sounds obvious but I arrived recently on the project and didn't know all theses concepts and how our projects were build into.

**Reply by neotrow** [reply]
@MerlinMason 

Did you ever find a solution for this? I'm in the same boat :-/ 

`useFormContext` returns null on SSR in nextjs...

**Reply by lasharela** [reply]
Any Solutions for NextJS 13?

**Reply by iamevansobeng** [reply]
Are you using a ui library?

**Reply by stevenirby** [reply]
@neotrow I too ran into an issue using `useFormContext` with next.js.

For now I just did this:

```
const { register, unregister, formState } = useFormContext() ?? {};
...

if (!register) return null;
```

Very dirty but I was able to build.

**Reply by neotrow** [reply]
@stevenirby 

That was also my first approach but a lot of our code is dependent on what we get back from `useFormContext` so it would just fail somewhere else.

My use case was that we created a package that contains some common components. We fixed it by exporting the `useFormContext` in our package:
https://github.com/neolution-ch/react-hook-form-components/blob/main/src/index.ts#L13

And then in the app itself importing the `useFormContext` from our package instead of from `react-hook-form-components` directly.

I think it's an issue with react contexts in general that get messy if they get initialised once in the app and once in the package but i didn't investigate any further.

**Reply by damneddutch** [reply]
@neotrow , absolute rockstar!  with a little tweaking of your code, got my shared component working as well!  What got me at first was that the component must reside inside the imported FormContext.Provider as well.

**Reply by hoa-nguyen-bk** [reply]
Thank you @stevenirby , because of little dirty code:
```
const { register, unregister, formState } = useFormContext() ?? {};
...

if (!register) return null;
```
 but I soon really that I forget the ```<FormProvider />```, you save my day, thank you so much

**Reply by paolobiavati** [reply]
> basically what i need it is just adding react-hook-form as external library in [webpack config](https://github.com/evgvis/test-lib/blob/main/webpack.config.js)

the linked git hub repo is private...

**Reply by DanMello** [reply]
For anyone running into this issue particularly in next js here is how I got it to work for me. Turns out the lib I made that uses react-hook-form was importing react-hook-form from its own node_modules folder instead of the next js projects node_modules folder. I am using typescript project references to include my lib code in next js so that could be part of the issue for me.

```
const nextConfig: NextConfig = {
  webpack: (config) => {
    config.resolve.alias["react-hook-form"] = path.resolve(
      __dirname,
      "node_modules/react-hook-form"
    );
    return config;
  },
};
```

**Reply by NNurtilek** [reply]
I faced this issue when using `useFormContext` inside a submodule package.
If you run into the same problem, just use `npm workspaces` to ensure that submodules use libraries from the root `node_modules`.

[a comment by an automated account is not shown]

**Comment by mcgrealc** [comment]
Stuck on this issue as well - using RHF 7.39.1, have added RHF to webpack externals, wrapped the component from my library in FormProvider but still useFormContext is null

**Reply by mcgrealc** [reply]
Just to answer my own question and to verify what @smeng9 said earlier in the thread - use rollup rather than webpack. My React App is ESM so rollup is the only option (only option that worked for me) to build your library to be compatible.

**Comment by thedocbwarren** [comment]
Can we open as a issue for better tracking?

**Comment by carboni-rob** [comment]
I think the easiest solution in this case is not to use the form context and apply a control inversion pattern. We need to pass the methods to the parent from the child, instead of getting them in the child via context. So we need something like:
```ts
const Form = ({formMethods}) => {
    return (
        <form onSubmit={formMethods.handleSubmit(onSubmit)}>
            {children}
            <Button .../>
        </form>
    );
};
```
and
```ts
const Input = (props) => {
   const methods = useForm();

    return (
        <Form
           formMethods={methods}
        >
           <CustomInput ... />
        </Form>
    )
}
```
This works for us, I hope it helps.

**Reply by alvamanu** [reply]
this kills the purpose of context though. You can simply destructure all methods you need from useForm when initializing it and then pass them as props

**Comment by UnicornRider26** [comment]
Beware of the different package versions in case this happens to you in cross-package usage situation (as it did to me)

**Reply by klapec** [reply]
I wish I found this comment sooner... Thank you!!!

**Comment by lhotamir** [comment]
If you are using rollup for your library and webpack for your app, make sure your rollup output is in the es format.

```
output: {
      file: pkg.main,
      format: 'es',
      ...
```

**Comment by jahabeebs** [comment]
For anyone reading this who happens to be using a component library that uses react-hook-form like shadcn/ui make sure your Form component is being imported from the component library you're using (like ```import { Form } from '@/components/ui/form'```) and not from react-hook-form itself!

**Reply by OwaisSiddiqui** [reply]
"Add all missing imports" 😬

**Reply by tika** [reply]
Legend, thank you!

**Reply by MichaelLegg** [reply]
Thanks, that was it for me too!

**Reply by zf0x00** [reply]
thanks mate

**Reply by SuHlaingHnin** [reply]
Thanks buddy

**Reply by qpages** [reply]
after 4 hours of struggle I read your comment, thanks 🙏

**Reply by MatheusBueno** [reply]
Work for me too! Thanks

**Reply by fhub-1** [reply]
thanks a lot working fine :((((

**Reply by codewithmecoder** [reply]
I did not expect that. thank you.

**Reply by altanyenigun** [reply]
this solved my problem too, thank you.

**Reply by mestir** [reply]
T1000

**Reply by ludo237** [reply]
Stumble upon this stupid import today, thanks

**Reply by monaabishek2003** [reply]
saved me bro,

**Reply by ArthurTimofey** [reply]
Doh!

**Comment by GlennOlsson** [comment]
For future reference;

I hade the same problem when creating a library that applies the context, when the projects using the library were Webpack Module Federated. It was not enough to only add `react-hook-form` as an external for my library, I also had to add it to the `shared` object of the `ModuleFederationPlugin` configuration, like 
```js
new ModuleFederationPlugin({
  shared: {
    "react-hook-form": {
      singleton: true,
      version: "7.27.0"
    }
    // ...
  },
  // ...
})
```

Hope this can help someone

**Comment by alvamanu** [comment]
I'd like to share my take on this.

Looks like as soon as you initiate your form with useForm, hooks from RHF library are inaccessible to children components. So I tried [sazzadtanim]s solution, but I'm not initiating props in children components before the return (all passed props or "methods" are run AFTER return). I'm actually running the prop directly and also making sure it runs only when it's accessible (props.getValues && props.getValues()). Looks like NextJS isn't following ReactJS loading patterns on build. Thank you sazzadtanim!

**Comment by edwardnz2017** [comment]
I also had this issue and found a simple workaround.

Original issue: 

```
export const Component = () => {
  // This is where the issue happened for me, as useFormContext was null
  const { setValue } = useFormContext()
  
  useEffect(() => {
    setValue("foo") 
  }, [setValue])
};
```

Fix:

```
export const Component = () => {
  const methods = useFormContext()
  
  useEffect(() => {
    if (methods && methods.setValue) {
      setValue("foo") 
    }
  }, [methods])
};
```

By simply adding a check to make sure that it isn't null, it got rid of the error.

**Comment by imranDev20** [comment]
const methods = useFormContext<ProductComboFormInput>() ?? {};

  const { fields, append, remove } = useFieldArray({
    control: methods && methods.control,
    name: "products",
  });
  
  
  Doesn't seem to be working for me.

**Comment by roaugusto** [comment]
Based on @edwardnz2017 comment, I was able to solve the problem in a simpler way:

```
type CustomInputProps = ComponentProps<'input'> & {
   name:string
}

export function CustomInput({ ...props }: CustomInputProps) {
   const methods = useFormContext()

   return <input id={props.name} {...props} {...methods?.register(props.name)} />
}
```

**Comment by Kepron** [comment]
If none of the other comments seem to match your situation, and you think everything in your code is correct and in right place, there's something I'd like you to check. It's pretty simple. If you render the function/const where you use `useFormContext` as a function, you will also get this error.

All I did was change this part of my code;

```
{filteredInputs.map(([key, items], index) => (
  <Grid key={key} size={{ xs: 12, sm: filteredInputs.length % 2 !== 0 && index === filteredInputs.length - 1 ? 12 : 6 }}>
    {ResAddRenderInputs(key, items)}
  </Grid>
))}
```

to this:

```
{filteredInputs.map(([key, items], index) => (
  <Grid key={key} size={{ xs: 12, sm: filteredInputs.length % 2 !== 0 && index === filteredInputs.length - 1 ? 12 : 6 }}>
    <ResAddRenderInputs itemKey={key} items={items} />
  </Grid>
))}
```

**Comment by acbonnyac** [comment]
Another option I'd add is to have your React component to take `useFormContext` as a prop. Something like this:
```tsx
interface Props {
    // useFormContext does not work when used across packages
    // We need to get this from the caller instead
    useFormContext?: typeof useFormContext;
}

const YourComponent: React.FC<Props> = ({
    useFormContext: useFormContextFunc = useFormContext,
}) => {
    const {
        watch,
        register,
        formState: { errors },
    } = useFormContextFunc();
    ...
};
```

This way, you can still test your component within NPM package by using `useFormContext` on its own project without having to pass around form methods. For the caller side on the other project, you can pass in `useFormContext` function to your React component so it's still going to work with `useForm` from outside.

**Comment by LazzyLizzard** [comment]
As of year 2025 and version 7.63.0 as a new user of this package I'm experiencing problem with null as return value. Quite unexpected I must say. Any plans to fix it without any workarounds?
